src/components/tl/mlx5/tl_mlx5_pd.c

```diff
@@ -0,0 +1,293 @@
+    }
+
+    addr             = (struct sockaddr_un *)&storage;
+    addr->sun_family = AF_UNIX;
+    strncpy(addr->sun_path, sock_path, sizeof(addr->sun_path));
+    addr->sun_path[sizeof(addr->sun_path) - 1] = '\0';
+
+    if (bind(sock, (struct sockaddr *)addr, sizeof(struct sockaddr_un)) == -1) {
+        tl_error(ctx->super.super.lib, "failed to bind server socket errno %d",
+                 errno);
+        return UCC_ERR_NO_MESSAGE;
+    }
+    if (listen(sock, group_size) == -1) {
+        tl_error(ctx->super.super.lib,
+                 "failed to listen to server socket errno %d", errno);
+        return UCC_ERR_NO_MESSAGE;
+    }
+    *sock_p = sock;
+    return UCC_OK;
+}
+
+static ucc_status_t client_recv_data(int *              shared_cmd_fd,
+                                     uint32_t *         shared_pd_handle,
+                                     const char *       sock_path,
+                                     ucc_tl_mlx5_lib_t *lib)
+{
+    struct sockaddr_storage sockaddr = {};
+    struct sockaddr_un *    addr;
+    int                     sock;
+    sock = socket(PF_LOCAL, SOCK_STREAM, 0);
+    if (sock == -1) {
+        tl_error(lib, "failed to create client socket errno %d", errno);
+        return UCC_ERR_NO_MESSAGE;
+    }
+
+    addr             = (struct sockaddr_un *)&sockaddr;
+    addr->sun_family = AF_UNIX;
+    strncpy(addr->sun_path, sock_path, sizeof(addr->sun_path));
+    addr->sun_path[sizeof(addr->sun_path) - 1] = '\0';
+
+    while (connect(sock, (struct sockaddr *)addr, SUN_LEN(addr)) == -1) {
+        if (errno != ENOENT) {
+            tl_error(lib, "failed to connect client socket errno %d", errno);
+            goto fail;
+        }
+    }
+    if (do_recvmsg(sock, shared_cmd_fd, shared_pd_handle) != UCC_OK) {
+        tl_error(lib, "Failed to recv msg");
+        goto fail;
+    }
+
+    if (close(sock) == -1) {
+        tl_error(lib, "Failed to close client socket errno %d", errno);
+        return UCC_ERR_NO_MESSAGE;
+    }
+    return UCC_OK;
+
+fail:
+    if (close(sock) == -1) {
+        tl_error(lib, "Failed to close client socket errno %d", errno);
+    }
+    return UCC_ERR_NO_MESSAGE;
+}
+
+static ucc_status_t server_send_data(int command_fd, uint32_t pd_handle,
+                                     int group_size, int sock,
+                                     ucc_tl_mlx5_lib_t *lib)
+{
+    int                i = 0;
+    connection_t       connection[group_size];
+    pthread_t          thread[group_size];
+    struct sockaddr_un addr;
+    socklen_t          addrlen;
+
+    while (i < group_size) {
+        /* accept incoming connections */
+        connection[i].sock      = accept(sock, NULL, 0);
+        connection[i].fd        = command_fd;
+        connection[i].pd_handle = pd_handle;
+        if (connection[i].sock != -1) {
+            /* start a new thread but do not wait for it */
+            pthread_create(&thread[i], 0, do_sendmsg, (void *)&connection[i]);
```

Comment:
My guess is that it is a way to handle the multiple connections in parallel.
But honestly, I don't know if there is a more specific reason. @vspetrov what do you think ?

Reply:
I believe it could be that i took threads from the "pd sharing example" that Rami gave me long ago. I don't see a scenario whith the dead-lock. I gues it should work just as well if server sends msg to peers one after another.

Reply:
thanks @vspetrov . @samnordmann  can you check if it works ok w/o threads here?